### ISSUE TITLE
Element “area” is missing required attribute “href”

### ISSUE THREAD

**Issue by Mintoo200** [validator/validator#1622]
URL being validated or code to reproduce error:
```html
<!DOCTYPE html>
<html lang="">
<head>
    <title>Test</title>
</head>
<body>

  <map name="infographic">
      <area shape="poly" nohref="" coords="129,0,260,95,129,138" alt="HTTP">
  </map>
  <img id="info" usemap="#infographic" src="/media/examples/mdn-info.png" alt="MDN infographic">

</body>
</html>
```

Produces the following errors:
```
"file:./scratch_115.html":8.5-8.74: error: Element “area” is missing required attribute “href”.
"file:./scratch_115.html":8.5-8.74: error: The “nohref” attribute on the “area” element is obsolete. Omitting the “href” attribute is sufficient.
```

The [`nohref` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/area#deprecated_attributes) is indeed deprecated but the `href` attribute is optional:

> href
The hyperlink target for the area. Its value is a valid URL. This attribute may be omitted; if so, the <area> element does not represent a hyperlink.

-- [MDN - `<area>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/area#attributes)

> The [href](https://html.spec.whatwg.org/multipage/links.html#attr-hyperlink-href) attribute on [a](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element) and [area](https://html.spec.whatwg.org/multipage/image-maps.html#the-area-element) elements is not required; when those elements do not have [href](https://html.spec.whatwg.org/multipage/links.html#attr-hyperlink-href) attributes they do not create hyperlinks.

-- [WHATWG - `href`](https://html.spec.whatwg.org/multipage/links.html#attr-hyperlink-href)

Note that the error about `href` being required still occurs when `nohref` is not present

**Comment by sideshowbarker** [comment]
Either remove the `alt` attribute from that `area` element, or else add an `href` attribute to the element — to make the markup snippet in the issue description conform to the spec requirements.

For the spec requirements, see https://html.spec.whatwg.org/multipage/image-maps.html#the-area-element:attr-hyperlink-href-4:

> If the [area](https://html.spec.whatwg.org/multipage/image-maps.html#the-area-element) element has no [href](https://html.spec.whatwg.org/multipage/links.html#attr-hyperlink-href) attribute, then the area represented by the element cannot be selected, **and the [alt](https://html.spec.whatwg.org/multipage/image-maps.html#attr-area-alt) attribute must be omitted**.

So, because the `alt` attribute is _not_ omitted for the `area` element in the markup snippet in the issue description, that `area` element is not allowed to lack an `href` attribute.

Thus it’s necessary to either omit the `alt` attribute or else to add an `href` attibute.

**Comment by Mintoo200** [comment]
Right, didn't know that. Then would it be possible to make the message more explicit ?
Something like "Element “area” is missing required attribute “href” when alt is provided" ?
I don't know if that is the only attribute making `href` required though

**Comment by sideshowbarker** [comment]
> would it be possible to make the message more explicit ?

It’s possible, but it requires taking some logic out of the portable (RelaxNG) schema/grammar that the checker uses, and moving it instead into some custom Java code. I’ve done it before, but it’s something that needs to be considered case-by-case, to decide if the specific case merits doing it, or not. Taken to a logical extreme, if we did it for everything, it would have the effect of turning the checker solely into a bunch of Java code, with no formalism at all to express any grammar for HTML in portable way that can potentially be used by other tools.

More specifically, what I mean is, the difference between what’s stored in this repo in the https://github.com/validator/validator/tree/main/schema/html5 tree, compared to what’s in https://github.com/validator/validator/blob/main/src/nu/validator/checker/schematronequiv/Assertions.java file.

And even more specifically, it’s about implementing a check such as what’s at: https://github.com/validator/validator/blob/65dc6939c470eeeaac0f40dddf52f099eded28dd/src/nu/validator/checker/schematronequiv/Assertions.java#L2570-L2571 …versus what’s at: https://github.com/validator/validator/blob/65dc6939c470eeeaac0f40dddf52f099eded28dd/schema/html5/embed.rnc#L583-L586

**Comment by Mintoo200** [comment]
Right because being an automata, it knows where it failed not how it arrived there right ? So when it reaches this section : 
```
		&	(	area.attrs.alt
>			&	shared-hyperlink.attrs.href
			)?
```
it knows it expected an `href`, not that it could have ignored that requirement if the whole block was not applicable.

**Comment by sideshowbarker** [comment]
> Right because being an automata, it knows where it failed not how it arrived there right ? So when it reaches this section :
> 
> ```
> 		&	(	area.attrs.alt
> 	>		&	shared-hyperlink.attrs.href
> 			)?
> ```
> 
> it knows it expected an `href`, not that it could have ignored that requirement if the whole block was not applicable.

Yeah, exactly that

**Comment by Mintoo200** [comment]
Just a thought, if we swap the 2 lines quoted to 
```
		&	(	shared-hyperlink.attrs.href
			&	area.attrs.alt
			)?
```
Then, if I am not mistaken, we would get the following:
```html
<!-- neither alt nor href -->
<area shape="poly" coords="129,0,260,95,129,138">
<!-- no href => does not enter the block => no error ✅ -->

<!-- no alt but href -->
<area shape="poly" href="#" coords="129,0,260,95,129,138">
<!-- href => enters the block => no alt => error ✅ -->
<!-- error would be something like "<area> must have an alt" -->

<!-- alt but no href -->
<area shape="poly" alt="test" coords="129,0,260,95,129,138">
<!-- no href => does not enter the block => alt is not part of the definition => error ✅ -->
<!-- error would be something like "<area> cannot have an alt" -->

<!-- both alt and href -->
<area shape="poly" alt="test" href="#" coords="129,0,260,95,129,138">
<!-- href => enters the block => alt present => no error ✅ -->
```
Wouldn't we ?

I would be happy to open a Pull Request with these changes if that reasoning makes sense.

**Comment by sideshowbarker** [comment]
> Just a thought, if we swap the 2 lines quoted

Swapping the lines won’t make any difference. The order doesn’t matter

**Comment by sideshowbarker** [comment]
I added some code to make the checker emit a new Info message:

> _Info: Either remove the `alt` attribute from this `area` element , or else, add an `href` attribute._

…and deployed the change:

https://validator.w3.org/nu/?showsource=yes&doc=data%3Atext%2Fhtml%3Bcharset%3Dutf-8%3Bbase64%2CPCFkb2N0eXBlIGh0bWw%2BPGh0bWwgbGFuZz48bWV0YSBjaGFyc2V0PXV0Zi04Pg0KPHRpdGxlPnRlc3Q8L3RpdGxlPg0KPGFyZWEgc2hhcGU9InBvbHkiIGNvb3Jkcz0iMTI5LDAsMjYwLDk1LDEyOSwxMzgiIGFsdD0iSFRUUCI%2B
